time-manager/System/TimeManager.hs

```diff
@@ -96,6 +100,22 @@ killManager = reaperKill
 
 ----------------------------------------------------------------
 
+-- | Registering a timeout action and unregister its handle
+--   when the body action is finished.
+withHandle :: Manager -> TimeoutAction -> (Handle -> IO a) -> IO a
+withHandle mgr onTimeout action =
+    E.bracket (register mgr onTimeout) cancel action
+
+-- | Registering a timeout action of killing this thread and
+--   unregister its handle when the body action is killed or finished.
+withHandleKillThread :: Manager -> TimeoutAction -> (Handle -> IO ()) -> IO ()
+withHandleKillThread mgr onTimeout action =
+    E.handle handler $ E.bracket (registerKillThread mgr onTimeout) cancel action
```

Comment:
Would it be better/worse to do the following? :thinking: 

    E.bracket (registerKillThread mgr onTimeout) cancel $ E.handle handler action

Reply:
Interesting.
I don't know which is better.

Reply:
I don't think it will matter much... since the `cancel` will be run either way.

The difference is this, effectively:

```hs
-- Here's the body of 'bracket'

-- either here instead of:
-- mask $ \restore -> do
E.handle handler $ mask $ \restore -> do
  a <- before
  -- or here instead of:
  -- r <- restore (thing a) `onException` after a
  r <- restore (E.handle handler $ thing a) `onException` after a
  _ <- after a
  return r
```